### PR TITLE
Added backend implementation for Resiliency score calculation and Total experiments passed calculation.

### PR DESCRIPTION
**Pull request by Jonsy13**
## Proposed changes

Summarize your changes here to communicate with the maintainers and make sure to put the link of that issue

**This PR will add backend implementation for-** 
- Resiliency score calculation 
- Total Experiments / Total Experiments passed

## Types of changes

What types of changes does your code introduce to Litmus? Put an `x` in the boxes that apply
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices applies)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
- [X] I have read the [CONTRIBUTING](https://github.com/litmuschaos/litmus/blob/master/CONTRIBUTING.md) doc
- [X] I have signed the commit for DCO to be passed.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added necessary documentation (if appropriate)

## Dependency
- Please add the links to the dependent PR need to be merged before this (if any).

## Special notes for your reviewer:

Sample Execution Data 
`
{"resiliency_score":"100","experiments_passed":"1","total_experiments":"1","event_type":"UPDATE","namespace":"litmus","name":"node-memory-hog-1617214746","creationTimestamp":"1617214754","phase":"Succeeded","message":"","startedAt":"1617214754","finishedAt":"1617215585","nodes":{"node-memory-hog-1617214746":{"name":"node-memory-hog-1617214746","phase":"Succeeded","message":"","startedAt":"1617214754","finishedAt":"1617215585","children":["node-memory-hog-1617214746-3029157648"],"type":"Steps"},"node-memory-hog-1617214746-1426460535":{"name":"install-experiment","phase":"Succeeded","message":"","startedAt":"1617214754","finishedAt":"1617215110","children":["node-memory-hog-1617214746-2022353413"],"type":"Pod"},"node-memory-hog-1617214746-2022206318":{"name":"[2]","phase":"Succeeded","message":"","startedAt":"1617215564","finishedAt":"1617215585","children":["node-memory-hog-1617214746-3233805379"],"type":"StepGroup"},"node-memory-hog-1617214746-2022353413":{"name":"[1]","phase":"Succeeded","message":"","startedAt":"1617215112","finishedAt":"1617215564","children":["node-memory-hog-1617214746-2938098115"],"type":"StepGroup"},"node-memory-hog-1617214746-2938098115":{"name":"run-chaos","phase":"Succeeded","message":"","startedAt":"1617215112","finishedAt":"1617215564","children":["node-memory-hog-1617214746-2022206318"],"type":"ChaosEngine","chaosData":{"engineUID":"b9f9fcf6-9eef-47ff-bbec-6ac88ced6515","engineName":"kube-proxy-node-memory-hog-chaos","namespace":"litmus","experimentName":"node-memory-hog","experimentStatus":"Completed","lastUpdatedAt":"1617215524","experimentVerdict":"Pass","experimentPod":"node-memory-hog-2l4y6l-x6pj9","runnerPod":"kube-proxy-node-memory-hog-chaos-runner","probeSuccessPercentage":"100","failStep":"N/A","chaosResult":{"metadata":{"name":"kube-proxy-node-memory-hog-chaos-node-memory-hog","namespace":"litmus","selfLink":"/apis/litmuschaos.io/v1alpha1/namespaces/litmus/chaosresults/kube-proxy-node-memory-hog-chaos-node-memory-hog","uid":"42db7486-5af5-4437-99fd-d85f4394e6b2","resourceVersion":"12029","generation":2,"creationTimestamp":"2021-03-31T18:30:34Z","labels":{"chaosUID":"b9f9fcf6-9eef-47ff-bbec-6ac88ced6515","controller-uid":"f0b962ca-9411-492b-9ede-dc81f5b6335d","job-name":"node-memory-hog-2l4y6l","name":"kube-proxy-node-memory-hog-chaos-node-memory-hog"},"managedFields":[{"manager":"experiments","operation":"Update","apiVersion":"litmuschaos.io/v1alpha1","time":"2021-03-31T18:31:53Z","fieldsType":"FieldsV1","fieldsV1":{"f:metadata":{"f:labels":{".":{},"f:chaosUID":{},"f:controller-uid":{},"f:job-name":{},"f:name":{}}},"f:spec":{".":{},"f:engine":{},"f:experiment":{}},"f:status":{".":{},"f:experimentstatus":{},"f:history":{}}}}]},"spec":{"engine":"kube-proxy-node-memory-hog-chaos","experiment":"node-memory-hog"},"status":{"experimentstatus":{"phase":"Completed","verdict":"Pass","failStep":"N/A","probeSuccessPercentage":"100"}}}}},"node-memory-hog-1617214746-3029157648":{"name":"[0]","phase":"Succeeded","message":"","startedAt":"1617214754","finishedAt":"1617215112","children":["node-memory-hog-1617214746-1426460535"],"type":"StepGroup"},"node-memory-hog-1617214746-3233805379":{"name":"revert-chaos","phase":"Succeeded","message":"","startedAt":"1617215564","finishedAt":"1617215583","children":null,"type":"Pod"}}}
`

### CODE REVIEW

File: litmus-portal/graphql-server/pkg/chaos-workflow/handler/handler.go [hunk 1]
```diff
@@ -272,11 +272,17 @@ func WorkFlowRunHandler(input model.WorkflowRunInput, r store.StateData) (string
 		return "", err
 	}
 
+	var executionData string = input.ExecutionData
+	// Resiliency Score will be calculated only if workflow execution is completed
+	if input.Completed {
+		executionData = ops.ResiliencyScoreCalculator(input.ExecutionData, input.WorkflowID)
+	}
```
Comment (gdsoumya) [inline litmus-portal/graphql-server/pkg/chaos-workflow/handler/handler.go:279]: Use `input.ExecutionData`

**Review by gdsoumya — CHANGES_REQUESTED** [review]

**Review by ishangupta-ds — CHANGES_REQUESTED** [review]
Since we are refactoring this and parsing the data on completion, can we also add the chaos engine status, chaos result verdict, probe success percentage, per experiment in an array of experiment results at workflow run level, instead of sending it as a chaos data in node to frontend @Jonsy13 it’s required for analytics.

**Comment by ishangupta-ds** [comment]
> Since we are refactoring this, can we also add the chaos engine status, chaos result verdict, probe success percentage, per experiment in an array of experiment results at workflow run level, instead of sending it as a chaos data in node to frontend @Jonsy13 it’s required for analytics.

File: litmus-portal/graphql-server/pkg/chaos-workflow/ops/operations.go [hunk 2]
```diff
@@ -329,3 +371,41 @@ func SendWorkflowEvent(wfRun model.WorkflowRun, r *store.StateData) {
...
+func ResiliencyScoreCalculator(execData string, wfid string) string {
+	var resiliency_score, weightSum, totalTestResult, totalExperiments, totalExperimentsPassed int = 0, 0, 0, 0, 0
+	var jsonData WorkflowEvent
+	json.Unmarshal([]byte(execData), &jsonData)
+	for _, value := range jsonData.Nodes {
+		if value.Type == "ChaosEngine" {
+			if value.ChaosExp == nil {
+				continue
+			}
+			chaosWorkflows, _ := dbOperationsWorkflow.GetWorkflows(bson.D{{"workflow_id", bson.M{"$in": []string{wfid}}}})
```
Comment (gdsoumya) [inline litmus-portal/graphql-server/pkg/chaos-workflow/ops/operations.go:385]: Move this outside loop as wfid will remain same, currently the db is queried multiple times for the same data.

**Review by gdsoumya — CHANGES_REQUESTED** [review]

**Comment by Jonsy13** [comment]
> Since we are refactoring this and parsing the data on completion, can we also add the chaos engine status, chaos result verdict, probe success percentage, per experiment in an array of experiment results at workflow run level, instead of sending it as a chaos data in node to frontend @Jonsy13 it’s required for analytics.

@ishangupta-ds Please create an issue to track this and modify the function according to your needs.
Currently, 2 teams are blocked due to this in frontend, so it can't be delayed.

File: litmus-portal/graphql-server/pkg/chaos-workflow/ops/operations.go (hunk 2 above)
Comment (gdsoumya) [reply litmus-portal/graphql-server/pkg/chaos-workflow/ops/operations.go:385]: wfid is unique so only 1 record will be returned. Directly use `chaosWorkflows[0]`

File: litmus-portal/graphql-server/pkg/chaos-workflow/ops/operations.go [hunk 3]
```diff
@@ -329,3 +371,41 @@ func SendWorkflowEvent(wfRun model.WorkflowRun, r *store.StateData) {
...
+			for _, workflow := range chaosWorkflows {
+				var Weightages []*model.Weightages
+				copier.Copy(&Weightages, &workflow.Weightages)
```
Comment (gdsoumya) [inline litmus-portal/graphql-server/pkg/chaos-workflow/ops/operations.go:388]: Not needed

**Review by gdsoumya — CHANGES_REQUESTED** [review]

File: litmus-portal/graphql-server/pkg/chaos-workflow/ops/operations.go [hunk 4]
```diff
@@ -329,3 +371,41 @@ func SendWorkflowEvent(wfRun model.WorkflowRun, r *store.StateData) {
...
+				for _, weightEntry := range Weightages {
+					if weightEntry.ExperimentName == value.ChaosExp.ExperimentName {
+						x, _ := strconv.Atoi(value.ChaosExp.ProbeSuccessPercentage)
+						totalTestResult += weightEntry.Weightage * x
+						weightSum += weightEntry.Weightage
+					}
+					if value.ChaosExp.ExperimentVerdict == "Pass" {
+						totalExperimentsPassed += 1
+					}
+				}
```
Comment (gdsoumya) [inline litmus-portal/graphql-server/pkg/chaos-workflow/ops/operations.go:399]: This loop can be moved out too, you can create a map of the probe success % and then outside the workflow node loop run the weightage loop and find the result required by mapping the exp names

**Review by gdsoumya — CHANGES_REQUESTED** [review]

**Review by gdsoumya — APPROVED** [review]
LGTM

**Review by vanshBhatia-A4k9 — APPROVED** [review]

**Pull request merged**